modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTableCommon.java

```diff
@@ -392,7 +393,13 @@ public static CompletableFuture<TableViewInternal> readTableAsync(ClientMessageU
      * @param tx The transaction.
      */
     public static void writeTxMeta(ClientMessagePacker out, @Nullable ClockService clockService, InternalTransaction tx) {
-        if (tx.remote()) {
+        if (out.resourceId() != 0) {
+            // Resource id is assigned on a first request in direct mode.
+            out.packLong(out.resourceId());
```

Comment:
Protocol change without a feature flag?

Reply:
It's safe here. This code path can only be executed by compatible client.